e2etest/l2tests/assignment.go

```diff
@@ -678,4 +697,267 @@ var _ = ginkgo.Describe("IP Assignment", func() {
+					Addresses: []string{
+						v6PoolAddresses,
+					},
+					AllocateTo: &metallbv1beta1.ServiceAllocation{Namespaces: []string{testNamespace}},
+				},
+			}
+			resources := config.Resources{
+				Pools: []metallbv1beta1.IPAddressPool{poolv4, poolv6},
+			}
+			err := ConfigUpdater.Update(resources)
+			Expect(err).NotTo(HaveOccurred())
+			ginkgo.By("Creating the service with PreferDualStack policy")
+			svc1, _ := service.CreateWithBackend(cs, testNamespace, "svc-test-ns-pool-2", func(svc *v1.Service) {
+				svc.Spec.IPFamilyPolicy = &IPFamilyPolicyPreferDualStack
+				svc.Spec.IPFamilies = []v1.IPFamily{v1.IPv6Protocol, v1.IPv4Protocol}
+			})
+			Expect(err).NotTo(HaveOccurred())
+
+			ginkgo.By("Validating that 1 IP is assigned to the service")
+			Eventually(func() int {
+				svc1, err = cs.CoreV1().Services(svc1.Namespace).Get(context.Background(), svc1.Name, metav1.GetOptions{})
+				Expect(err).NotTo(HaveOccurred())
+				return len(svc1.Status.LoadBalancer.Ingress)
+			}, 5*time.Minute, 1*time.Second).Should(Equal(1))
+			ginkgo.By("validate LoadBalancer IP is allocated from ipv6 pool")
+			err = config.ValidateIPInRange([]metallbv1beta1.IPAddressPool{poolv6}, jigservice.GetIngressPoint(
+				&svc1.Status.LoadBalancer.Ingress[0]))
+			Expect(err).NotTo(HaveOccurred())
+		})
+		ginkgo.It("Additional ip should be assigned when the assigned 1-stack pool becomes dual-stack", func() {
+			pool1 := metallbv1beta1.IPAddressPool{
+				ObjectMeta: metav1.ObjectMeta{Name: "test-ns-dualstack-pool"},
+				Spec: metallbv1beta1.IPAddressPoolSpec{
+					Addresses: []string{
+						v4PoolAddresses,
+					},
+					AllocateTo: &metallbv1beta1.ServiceAllocation{Namespaces: []string{testNamespace}},
+				},
+			}
+			resources := config.Resources{
+				Pools: []metallbv1beta1.IPAddressPool{pool1},
+			}
+
+			err := ConfigUpdater.Update(resources)
+			Expect(err).NotTo(HaveOccurred())
+
+			ginkgo.By("Creating the service with PreferDualStack policy")
+			svc1, _ := service.CreateWithBackend(cs, testNamespace, "svc-test-ns-pool-1", func(svc *v1.Service) {
+				svc.Spec.IPFamilyPolicy = &IPFamilyPolicyPreferDualStack
+			})
+			Expect(err).NotTo(HaveOccurred())
+
+			ginkgo.By("Validating that 1 IP is assigned to the service")
+			Eventually(func() int {
+				svc1, err = cs.CoreV1().Services(svc1.Namespace).Get(context.Background(), svc1.Name, metav1.GetOptions{})
+				Expect(err).NotTo(HaveOccurred())
+				return len(svc1.Status.LoadBalancer.Ingress)
+			}, 5*time.Minute, 1*time.Second).Should(Equal(1))
+			ginkgo.By("validate LoadBalancer IP is allocated from pool1")
+			err = config.ValidateIPInRange([]metallbv1beta1.IPAddressPool{pool1}, jigservice.GetIngressPoint(
+				&svc1.Status.LoadBalancer.Ingress[0]))
+			Expect(err).NotTo(HaveOccurred())
+			firstIP := svc1.Status.LoadBalancer.Ingress[0].String()
+			ginkgo.By("Updating pool1 to include both v4 and v6 addresses")
+			pool1.Spec.Addresses = []string{v4PoolAddresses, v6PoolAddresses}
```

Comment:
can we modify this to resemble what I commented on? that is, we also extend the ipv4 range (but still include the original ip), to be sure we aren't preserving the original ipv4 just because there is only a single one in the pool